### PR TITLE
[Pipelines] Load public gemma4_unified checkpoints (Gemma 4 12B) and serve them text-only

### PR DESCRIPTION
**Pull request by msaelices**
## Summary

Makes the public `gemma4_unified` checkpoints (the Gemma 4 12B line: `google/gemma-4-12b-it` and derived quants) loadable by MAX, serving them **text-only**. Fixes the first two blockers of #6669 and implements the text-only path proposed there.

- Register an `AutoConfig` shim for `model_type: "gemma4_unified"` (transformers < 5.5 path) and `gemma4_unified_arch`, a `dataclasses.replace` alias of `gemma4_arch` under the public `Gemma4UnifiedForConditionalGeneration` name — these checkpoints carry the regular Gemma 4 language layout with **no bundled MTP draft weights**, so they belong to `gemma4`, not `unified_mtp_gemma4`.
- Text-only serving: the unified checkpoints use a lightweight `vision_embedder` with a different schema than `Gemma4VisionModel`; detect it, set `vision_config=None`, skip building/loading the vision graph, and return 0 vision-cache bytes.
- `finalize` now passes the real `model.language_model.` prefixes to `parse_quant_config`, so per-layer quantized/ignored classification stops looking up `model.layers.*` keys that never exist (ignore-listed BF16 attention in modelopt 12B quants was never recognized as ignored).
- Honor `quant_config.attn_quantized_layers` per layer instead of assuming attention is always BF16 under NVFP4 (true for nvidia first-party checkpoints, false for the community 12B quants), and drop modelopt `k_scale`/`v_scale` KV-cache scales in the weight adapter.

Self-review (adversarial pass) caught and fixed an inference-path crash before any human review: `prepare_initial_token_inputs` read `vision_config.pooling_kernel_size` unconditionally, so text-only serving would have failed on the first prefill; image requests now get a clear rejection and the unified alias declares TEXT-only modalities.

## Verification (NVIDIA A10G 24 GB; bazel-built serve)

Each fix was driven by serving real checkpoints; the chain `config → arch resolution → graph build → weight load` now completes for `google/gemma-4-12b-it`-family checkpoints where each stage failed before (`gemma4_unified` not recognized → no matching architecture → vision `AttributeError` → o_proj shape mismatch → unexpected `k_scale`/`v_scale`).

**Full text-generation E2E verified on the A10G** (`berkerdooo/gemma-4-12B-it-NVFP4`, attention quantized): correct, coherent EN/ES completions via `/v1/chat/completions`. With #6668's fused dequant-GEMV (no BF16 weight materialization), the 12B sustains **31.5 tok/s warm decode at batch 1 on the 24 GB card** (measured on a branch combining this PR with #6668) — ~2x vLLM nightly's Marlin path on the same checkpoint and GPU — with token-identical output across the fallback's kernel revisions. Note: on a 24 GB card the stock activation-memory estimate (~15 GiB headroom, sized for the vision configs) still blocks startup — the measurement lowered it locally; right-sizing that estimate for text-only unified configs is intentionally left out of this PR pending maintainer guidance.

Related: #6669 (blocker 3 — implementing the unified `vision_embedder` — remains open), #6668, #6665.

Assisted-by: AI

### CODE REVIEW

File: max/python/max/pipelines/architectures/gemma4/weight_adapters.py [hunk 1]
```diff
@@ -44,6 +44,10 @@ def convert_safetensor_language_state_dict(
     new_state_dict: dict[str, WeightData] = {}
 
     for weight_name, value in state_dict.items():
+        # modelopt checkpoints may carry FP8 KV-cache scales (k_scale /
+        # v_scale) that MAX's BF16 KV cache does not consume; drop them.
+        if weight_name.endswith((".k_scale", ".v_scale")):
```
Comment (k-w-w) [inline max/python/max/pipelines/architectures/gemma4/weight_adapters.py:49]: Is the model still accurate even when these scales are dropped?

File: max/python/max/pipelines/architectures/gemma4/model_config.py [hunk 2]
```diff
@@ -651,6 +674,13 @@ def finalize(
             huggingface_config,
             state_dict,
             self.dtype,
+            # Gemma 4 checkpoints nest the language tower under
+            # "model.language_model."; without these prefixes the per-layer
+            # quantized/ignored classification looks up "model.layers.*"
+            # keys that never exist, so ignore-listed (BF16) attention in
+            # modelopt 12B quants was never recognized as ignored.
+            state_dict_name_prefix="model.language_model.",
+            ignored_modules_prefix="model.language_model.",
```
Comment (k-w-w) [inline max/python/max/pipelines/architectures/gemma4/model_config.py:686]: cc @callumcurtis could you double check that this change is compatible with existing quantized Gemma4 models?

File: max/python/max/pipelines/architectures/gemma4/weight_adapters.py (hunk 1 above)
Comment (msaelices) [reply max/python/max/pipelines/architectures/gemma4/weight_adapters.py:49]: Yes. `k_scale`/`v_scale` are modelopt's FP8 KV-cache quantization scales, but these checkpoints ship `kv_cache_quant_algo: null` (verified on berkerdooo/gemma-4-12B-it-NVFP4) and MAX uses a BF16 KV cache, so nothing consumes them, they're vestigial tensors. With strict=True loading, MAX errors on them as unexpected keys unless dropped. Decode output is token-identical with and without the drop at temp 0.

File: max/python/max/pipelines/architectures/gemma4/model_config.py (hunk 2 above)
Comment (msaelices) [reply max/python/max/pipelines/architectures/gemma4/model_config.py:686]: The prefixes match how modelopt's `ignored`/quantized module lists are written for the first-party checkpoints: nvidia/Gemma-4-26B-A4B-NVFP4 and -31B list their ignored attention under `model.language_model.*` globs (verified from their hf_quant_config). Before this change those globs never matched (we looked up `model.layers.*`), so ignore-listed BF16 attention was silently treated as quantized

File: max/python/max/pipelines/architectures/gemma4/gemma4.py [hunk 3]
```diff
@@ -134,6 +134,14 @@ def __init__(self, config: Gemma4ForConditionalGenerationConfig) -> None:
...
             is_nvfp4 = quant_config is not None and quant_config.is_nvfp4
             moe_nvfp4 = is_nvfp4 and text_config.enable_moe_block
+            # The first-party NVFP4 checkpoints (nvidia/Gemma-4-*) keep
+            # attention in BF16, but other modelopt quants (e.g. the
+            # community 12B NVFP4 ones) quantize it too -- honor the
+            # per-layer classification instead of assuming BF16.
+            attn_quantized = is_nvfp4 and (
+                quant_config is not None
+                and i in quant_config.attn_quantized_layers
+            )
```
[the remark this answers is not in the document]
Comment (msaelices) [reply max/python/max/pipelines/architectures/gemma4/gemma4.py]: Fixed, dropped the redundant `quant_config is not None` (is_nvfp4 implies it).

File: max/python/max/pipelines/architectures/gemma4/model_config.py [hunk 4]
```diff
@@ -440,7 +456,7 @@ class Gemma4ForConditionalGenerationConfig(ArchConfigWithKVCache):
     text_config: Gemma4TextConfig
     """The config object of the text backbone."""
 
-    vision_config: Gemma4VisionConfig
+    vision_config: Gemma4VisionConfig | None
     """The config object of the vision encoder."""
```
[the remark this answers is not in the document]
Comment (msaelices) [reply max/python/max/pipelines/architectures/gemma4/model_config.py]: Fixed — the vision_config field docstring now documents the text-only `None` case.

File: max/python/max/pipelines/architectures/gemma4/model_config.py [hunk 5]
```diff
@@ -593,9 +609,16 @@ def initialize_from_config(
...
+        vision_config: Gemma4VisionConfig | None
+        if hasattr(hf_vision_config, "hidden_activation"):
+            vision_config = Gemma4VisionConfig.initialize_from_config(
+                hf_vision_config
+            )
+        else:
+            # The "gemma4_unified" checkpoints (e.g. google/gemma-4-12b-it)
+            # use a lightweight vision_embedder with a different schema that
+            # is not implemented yet; serve text-only.
+            vision_config = None
```
[the remark this answers is not in the document]
Comment (msaelices) [reply max/python/max/pipelines/architectures/gemma4/model_config.py]: Fixed. Now gates on `model_type == "gemma4_unified"` instead of a missing attribute, so a malformed full-vision config fails loudly in initialize_from_config instead of silently degrading to text-only.

File: max/python/max/pipelines/architectures/gemma4/memory_planner.py [hunk 6]
```diff
@@ -91,6 +91,10 @@ def estimate_vision_cache_entry_bytes(
             raise ValueError(
                 "Gemma4 requires a text_config in the HuggingFace config"
             )
+        if not hasattr(vision_config, "position_embedding_size"):
+            # "gemma4_unified" checkpoints use a different vision schema and
+            # are served text-only; no vision cache is needed.
+            return 0
```
[the remark this answers is not in the document]
Comment (msaelices) [reply max/python/max/pipelines/architectures/gemma4/memory_planner.py]: Fixed, same model_type gate here as in model_config.

File: max/python/max/pipelines/architectures/gemma4/model.py [hunk 7]
```diff
@@ -646,6 +660,17 @@ def prepare_initial_token_inputs(
             if context_batch
             else False
         )
+        if needs_images and self.vision_model is None:
+            raise ValueError(
+                "This checkpoint is served text-only (its gemma4_unified"
+                " vision embedder is not implemented); image inputs are"
+                " not supported."
+            )
```
[the remark this answers is not in the document]
Comment (msaelices) [reply max/python/max/pipelines/architectures/gemma4/model.py:658]: Fixed, message is now model-agnostic ("no vision encoder is loaded").

File: max/python/max/pipelines/architectures/gemma4/model_config.py [hunk 8]
```diff
@@ -69,6 +69,22 @@ def __init__(
...
+    native Gemma4Config may not register the unified model_type.
+    """
+
+    model_type = "gemma4_unified"
+
+
+try:
+    AutoConfig.register("gemma4_unified", _Gemma4UnifiedHFConfig)
+except ValueError:
+    pass
```
[the remark this answers is not in the document]
Comment (msaelices) [reply max/python/max/pipelines/architectures/gemma4/model_config.py:85]: Fixed, same model-agnostic message.

File: max/python/max/pipelines/architectures/gemma4/model.py [hunk 9]
```diff
@@ -453,6 +463,12 @@ def _build_vision_graph(
         return vision_graph, vision_model.state_dict()
 
     def _run_vision_encoder(self, raw: VisionRawInputs) -> list[Buffer]:
+        if self.vision_model is None:
+            raise ValueError(
+                "This checkpoint is served text-only (its gemma4_unified"
+                " vision embedder is not implemented); image and video"
+                " inputs are not supported."
+            )
```
[the remark this answers is not in the document]
Comment (msaelices) [reply max/python/max/pipelines/architectures/gemma4/model.py:462]: Fixed, message is now model-agnostic ("no vision encoder is loaded").

File: max/python/max/pipelines/architectures/gemma4/model_config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (msaelices) [reply max/python/max/pipelines/architectures/gemma4/model_config.py:85]: Correcting my earlier reply (wrong thread): on the import side effect — kept module-level intentionally. The sibling `gemma4` registration right above (`AutoConfig.register("gemma4", ...)`) is already module-level under the same pattern, and the shim must exist before any `AutoConfig.from_pretrained` of a unified checkpoint; deferring to register_all_models() would risk that ordering. The `except` is already narrowed to `ValueError` (idempotent re-registration).

**Review by k-w-w — APPROVED** [review]
Thanks for confirming the accuracy 🙏

**Comment by k-w-w** [comment]
!sync

**Comment by k-w-w** [comment]
Note - I have a followup fix for the AutoConfig registration, which is needed for this model to run (model architectures are now loaded lazily, so the code that registers the new AutoConfig model types needs to be moved elsewhere).
